ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidatorFactory.scala

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.validator.batch
+
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{DamlStateKey, DamlStateValue}
+import com.daml.ledger.validator.LedgerStateOperations.{Key, Value}
+import com.daml.ledger.validator.{
+  CachingCommitStrategy,
+  CachingDamlLedgerStateReader,
+  CommitStrategy,
+  DamlLedgerStateReader,
+  DefaultStateKeySerializationStrategy,
+  LedgerStateOperations,
+  LedgerStateReader,
+  LogAppendingCommitStrategy,
+  QueryableReadSet,
+  StateKeySerializationStrategy
+}
+import com.daml.lf.engine.Engine
+import com.daml.metrics.Metrics
+import com.github.benmanes.caffeine.cache.Cache
+
+import scala.concurrent.{ExecutionContext, Future}
+
+object BatchedSubmissionValidatorFactory {
+  class LedgerStateReaderAdapter[LogResult](delegate: LedgerStateOperations[LogResult])
+      extends LedgerStateReader {
+    override def read(keys: Seq[Key]): Future[Seq[Option[Value]]] = delegate.readState(keys)
+  }
+
+  def readerAndCommitStrategyFrom[LogResult](
+      ledgerStateOperations: LedgerStateOperations[LogResult],
+      keySerializationStrategy: StateKeySerializationStrategy = DefaultStateKeySerializationStrategy)(
+      implicit executionContext: ExecutionContext)
+    : (DamlLedgerStateReader, CommitStrategy[LogResult]) = {
+    val ledgerStateReader = DamlLedgerStateReader.from(
+      new LedgerStateReaderAdapter[LogResult](ledgerStateOperations),
+      keySerializationStrategy)
+    val commitStrategy =
+      new LogAppendingCommitStrategy[LogResult](ledgerStateOperations, keySerializationStrategy)
+    (ledgerStateReader, commitStrategy)
+  }
+
+  def cachingReaderAndCommitStrategyFrom[LogResult](
+      ledgerStateOperations: LedgerStateOperations[LogResult],
+      stateCache: Cache[DamlStateKey, DamlStateValue],
+      keySerializationStrategy: StateKeySerializationStrategy = DefaultStateKeySerializationStrategy)(
+      implicit executionContext: ExecutionContext)
+    : (DamlLedgerStateReader with QueryableReadSet, CommitStrategy[LogResult]) = {
+    val ledgerStateReader = new CachingDamlLedgerStateReader(
+      stateCache,
```

Comment:
Prefer factory methods over `apply` in case multiple components need to be instantiated. Internal code (such as the above) uses constructors directly unless using `apply` has less maintenance.
Do you have any specific suggestions in mind?